src/terms/local_nonlinearity.jl

```diff
@@ -11,10 +11,10 @@ struct TermLocalNonlinearity{TF} <: TermNonlinear
 end
 (L::LocalNonlinearity)(::AbstractBasis) = TermLocalNonlinearity(L.f)
 
-function ene_ops(term::TermLocalNonlinearity, basis::PlaneWaveBasis, ψ, occ; ρ, kwargs...)
+function ene_ops(term::TermLocalNonlinearity, basis::PlaneWaveBasis{T}, ψ, occ; ρ, kwargs...) where {T}
     fp(ρ) = ForwardDiff.derivative(term.f, ρ)
     E = sum(term.f.(ρ)) * basis.dvol
-    potential = fp.(ρ)
+    potential = convert_dual.(T, fp.(ρ))
```

Comment:
it's not clear to me why that is needed?

Reply:
In case you have a dual in the function returning a potential. (like in the forwarddiff example)

Reply:
this is scary to me, how do you know you're not discarding important information? Also will we need to pepper the code with this? Let's discuss on a call

Reply:
The way out is to properly refactor the floating-point type handling as we do it now. I agree that this should be done. This is only the "quick way out".

You are not discarding important information, since this `convert_dual` is only triggered in case the type of the PlaneWaveBasis (which in our current setup overrules everything) is non-dual and thus it is sort of requested to do a non-dual computation. Pushing the dual through in that case you will anyway get inconsistent results as not all dual rules will be triggered.